OSX System.Net.NetworkInformation.Functional.Tests timeout in CI

http://dotnet-ci.cloudapp.net/job/dotnet_corefx/job/master/job/osx_debug_prtest/263/console

> 08:53:21   Discovering: System.Net.NetworkInformation.Functional.Tests
08:53:21   Discovered:  System.Net.NetworkInformation.Functional.Tests
08:53:21   Starting:    System.Net.NetworkInformation.Functional.Tests
08:53:21   System.Xml.XDocument.xNodeBuilder.Tests -> /Users/dotnet-bot/j/workspace/dotnet_corefx/master/osx_debug_prtest/bin/AnyOS.AnyCPU.Debug/System.Xml.XDocument.xNodeBuilder.Tests/System.Xml.XDocument.xNodeBuilder.Tests.dll
10:22:33 
Build timed out (after 120 minutes). Marking the build as aborted.
10:22:39 
Build was aborted
10:22:39 [xUnit] [INFO] - Starting to record.
10:22:39 [xUnit] [INFO] - Processing xUnit.Net-v2 (default)
10:22:39 
MSBUILD : error MSB4166: Child node "3" exited prematurely. Shutting down. Diagnostic information may be found in files in the temporary files directory named MSBuild_*.failure.txt.
10:22:39 /Users/dotnet-bot/j/workspace/dotnet_corefx/master/osx_debug_prtest/Tools/tests.targets(145,5): warning MSB5021: "sh" and its child processes are being terminated in order to cancel the build. [/Users/dotnet-bot/j/workspace/dotnet_corefx/master/osx_debug_prtest/src/System.Net.NetworkInformation/tests/FunctionalTests/System.Net.NetworkInformation.Functional.Tests.csproj]

--- a/src/Common/src/Interop/OSX/Interop.RunLoop.cs
+++ b/src/Common/src/Interop/OSX/Interop.RunLoop.cs
@@ -61,5 +61,16 @@ internal static partial class Interop
         /// <param name="rl">The run loop mode of rl from which to remove source.</param>
         [DllImport(Interop.Libraries.CoreFoundationLibrary)]
         internal static extern void CFRunLoopRemoveSource(CFRunLoopRef rl, CFRunLoopSourceRef source, CFStringRef mode);
+        
+        /// <summary>
+        /// Returns a bool that indicates whether the run loop is waiting for an event.
+        /// </summary>
+        /// <param name="rl">The run loop to examine.</param>
+        /// <returns>true if rl has no events to process and is blocking,
+        /// waiting for a source or timer to become ready to fire;
+        /// false if rl either is not running or is currently processing
+        /// a source, timer, or observer.</returns>
+        [DllImport(Interop.Libraries.CoreFoundationLibrary)]
+        internal static extern bool CFRunLoopIsWaiting(CFRunLoopRef rl);
     }
 }

--- a/src/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.OSX.cs
+++ b/src/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.OSX.cs
@@ -172,6 +172,9 @@ namespace System.Net.NetworkInformation
             Debug.Assert(s_runLoopSource != null);
             Debug.Assert(s_dynamicStoreRef != null);
 
+            // Allow RunLoop to finish current processing.
+            while (!Interop.RunLoop.CFRunLoopIsWaiting(s_runLoop)) { }
+
             Interop.RunLoop.CFRunLoopStop(s_runLoop);
             s_runLoopEndedEvent.WaitOne();
         }

--- a/src/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.OSX.cs
+++ b/src/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.OSX.cs
@@ -173,7 +173,7 @@ namespace System.Net.NetworkInformation
             Debug.Assert(s_dynamicStoreRef != null);
 
             // Allow RunLoop to finish current processing.
-            while (!Interop.RunLoop.CFRunLoopIsWaiting(s_runLoop)) { }
+            SpinWait.SpinUntil(() => Interop.RunLoop.CFRunLoopIsWaiting(s_runLoop));
 
             Interop.RunLoop.CFRunLoopStop(s_runLoop);
             s_runLoopEndedEvent.WaitOne();

--- a/src/System.Net.NetworkInformation/tests/FunctionalTests/NetworkChangeTest.cs
+++ b/src/System.Net.NetworkInformation/tests/FunctionalTests/NetworkChangeTest.cs
@@ -9,7 +9,6 @@ namespace System.Net.NetworkInformation.Tests
     public class NetworkChangeTest
     {
         [Fact]
-        [ActiveIssue(8066, PlatformID.OSX)]
         public void NetworkAddressChanged_AddRemove_Success()
         {
             NetworkAddressChangedEventHandler handler = NetworkChange_NetworkAddressChanged;

